plugins/processors/traffic_shaper/traffic_shaper.go

```diff
@@ -0,0 +1,83 @@
+package traffic_shaper
+
+import (
+	_ "embed"
+	"sync"
+	"time"
+
+	"github.com/influxdata/telegraf"
+	"github.com/influxdata/telegraf/internal/limiter"
+	"github.com/influxdata/telegraf/plugins/processors"
+	"github.com/influxdata/telegraf/selfstat"
+)
+
+//go:embed sample.conf
+var sampleConfig string
+
+type TrafficShaper struct {
+	Samples          int `toml:"samples"`
+	TimeUnit         time.Duration
+	BufferSize       int `toml:"buffer_size"`
+	Queue            chan *telegraf.Metric
+	Acc              telegraf.Accumulator
+	Log              telegraf.Logger `toml:"-"`
+	wg               sync.WaitGroup
+	messagesInFlight selfstat.Stat
+	messagesDropped  selfstat.Stat
+}
+
+func (*TrafficShaper) SampleConfig() string {
+	return sampleConfig
+}
+
+func (t *TrafficShaper) Start(acc telegraf.Accumulator) error {
+	t.Queue = make(chan *telegraf.Metric, t.BufferSize)
+	t.Acc = acc
+	t.wg.Add(1)
+	go t.ShapeTraffic()
+	t.messagesInFlight = selfstat.Register("traffic_shaper", "messages_inflight", map[string]string{})
+	t.messagesDropped = selfstat.Register("traffic_shaper", "messages_dropped", map[string]string{})
+	return nil
+}
+
+func init() {
+	processors.AddStreaming("traffic_shaper", func() telegraf.StreamingProcessor {
+		return newTrafficShaper()
+	})
+}
+
+func (t *TrafficShaper) Stop() {
+	t.Log.Debugf("Got stop signal %s", time.Now().String())
+	close(t.Queue)
+	t.wg.Wait()
```

Comment:
This will block telegraf from exiting until all metrics are flushed from the queue? Not sure this is the behavior we want. When someone closes or stops telegraf, things should clean up, but this could block for 100s or 1000s of seconds.

Reply:
added this as a config so that for users can choose accordingly.